### ISSUE TITLE
Lens Distortion Pass

### ISSUE THREAD

**Issue by eriksachse** [pmndrs/postprocessing#440]
### Is your feature request related to a problem?

Is there any interest in adding a lens distortion pass to pmndrs/postprocessing?

### Describe the solution you'd like

I made a custom pass that imitates a lens distortion as seen in camera lenses 🎥

### Additional context

<img width="1071" alt="Screenshot 2022-12-22 at 14 09 17" src="https://user-images.githubusercontent.com/24212703/209141221-d8b35cf0-3816-4912-9ed7-f8778ee85d14.png">

**Comment by vanruesc** [comment]
Hi,

what tasks would this pass perform? There's already the `ChromaticAberrationEffect` which supports color shifting near the screen edges and I think lens/perspective warping can be done by changing camera parameters :thinking:

**Comment by eriksachse** [comment]
I'm aware of all these parameters. 
I made a simple demo to showcase what I mean: [https://laens-1b1ohvvdz-eriksachse.vercel.app/ ](https://laens-1b1ohvvdz-eriksachse.vercel.app/ )
It'll add a barrel/fisheye and pincushion distortion. More info here: [https://learnopencv.com/understanding-lens-distortion/](https://learnopencv.com/understanding-lens-distortion/)
I'd have to tweak the parameters further to avoid the frustum culling glitch.

**Comment by vanruesc** [comment]
Feel free to send a PR.

**Comment by danrossi** [comment]
That project seems unusable, its built into an app. But I've been testing this out for my purpose. But its applying the whole viewport for both WebXR cameras  https://github.com/ycw/three-lens-distortion

This is what I can come up with so far, I cant make it distort each webxr camera and sized for each viewport. Trying to rebuild a webxr polyfill solution for Iphone as the polyfill project is broken with three. People want it. 

![Screenshot 2023-07-09 070510](https://github.com/pmndrs/postprocessing/assets/710277/409866c5-8f10-4fea-86ca-142ce86e8041)

**Comment by danrossi** [comment]
I have a working ported example. I'm not sure if it's correct. 

https://danrossi.github.io/three-lens-distortion/example/postprocessing.html

Shader. Takes inputBuffer as input to manipulate uv of the texture. 

https://github.com/danrossi/three-lens-distortion/blob/main/src/glsl/lens-distortion.frag

**Comment by vanruesc** [comment]
@danrossi Thanks for porting that effect. I think the `mainUv` function would be better suited in this case :thinking: Can you please check if the following works the same?

```js
uniform vec2 distortion;
uniform vec2 principalPoint; 
uniform vec2 focalLength; 
uniform float skew;

float border(const in vec2 uv) {

	return float(uv.s >= 0.0 && uv.s <= 1.0 && uv.t >= 0.0 && uv.t <= 1.0);

}

void mainUv(inout vec2 uv) {

	vec2 xn = 2.0 * (uv.st - 0.5); // [-1, 1]
	vec3 xDistorted = vec3((1.0 + distortion * dot(xn, xn)) * xn, 1.0);

	mat3 kk = mat3(
		vec3(focalLength.x, 0.0, 0.0),
		vec3(skew * focalLength.x, focalLength.y, 0.0),
		vec3(principalPoint.x, principalPoint.y, 1.0)
	);

	uv = (kk * xDistorted).xy * 0.5 + 0.5;

}

void mainImage(const in vec4 inputColor, const in vec2 uv, out vec4 outputColor) {

	outputColor = vec4(border(uv) * inputcolor.rgb, inputcolor.a);

}
```

**Comment by danrossi** [comment]
Yes that worked. I've got to get it working in XR now. It shows up doubling up over both cameras, in the html emulator and polyfill its not rendering a texture. Not sure if its a shader issue. 

https://danrossi.github.io/three-lens-distortion/example/postprocessing.html

**Comment by danrossi** [comment]
actually it doesn't cut around the texture. It's filled black in the canvas. Before it was showing the background color of the page.

**Comment by vanruesc** [comment]
> actually it doesn't cut around the texture

It's probably working as intended now. Try setting the clear alpha to 0: `renderer.setClearAlpha(0)`

**Comment by danrossi** [comment]
I've added it to this option. 

https://danrossi.github.io/three-lens-distortion/example/postprocessing-uv.html

the original

https://danrossi.github.io/three-lens-distortion/example/postprocessing.html

**Comment by vanruesc** [comment]
Hm, what if you change [this line](https://github.com/danrossi/three-lens-distortion/blob/08041ba3c602e94c03936c175f7967a4b67f1542/src/glsl/lens-distortion-uv.frag#L29) to `outputColor = border(uv) * inputcolor;`

**Comment by danrossi** [comment]
It worked. I added a new example with uniforms menu
https://danrossi.github.io/three-lens-distortion/example/postprocessing.html
https://danrossi.github.io/three-lens-distortion/example/postprocessing-video.html

**Comment by danrossi** [comment]
Should I add that as a PR ? I prefer raw tests however all the demos in this project are difficult to pull apart the code and figure out usage. It took me a while.

**Comment by vanruesc** [comment]
> Should I add that as a PR ?

Yes, that's fine. No need to add a demo for this. Here's the [PR guide](https://github.com/pmndrs/postprocessing/blob/main/.github/CONTRIBUTING.md#pull-request-process-sparkles), just in case.

FYI: the new demos are here: https://github.com/pmndrs/postprocessing/tree/main/manual/assets/js/src/demos

**Comment by danrossi** [comment]
I'll try my very best and see what I can do. I already made a fork but made changes to main for testing something. I'll set it up again to do changes in a branch.
